File: app/src/main/java/org/digitalcampus/oppia/analytics/CountlyAnalytics.java
package org.digitalcampus.oppia.analytics;

import android.app.Activity;
import android.content.Context;

import org.digitalcampus.mobile.learning.BuildConfig;

import java.util.HashMap;
import java.util.Map;

import ly.count.android.sdk.Countly;
import ly.count.android.sdk.CountlyConfig;
import ly.count.android.sdk.DeviceId;

public class CountlyAnalytics extends BaseAnalytics {

    public CountlyAnalytics(Context ctx) {
        super(ctx);
    }

    @Override
    protected void startTrackingSession() {
        CountlyConfig config = new CountlyConfig(ctx, BuildConfig.COUNTLY_APP_KEY, BuildConfig.COUNTLY_SERVER_URL);
        config.setLoggingEnabled(Analytics.isTrackingEnabled(ctx));
        config.setViewTracking(Analytics.isTrackingEnabled(ctx));
        if (Analytics.isBugReportEnabled(ctx)) {
            config.enableCrashReporting();
        }
        config.setIdMode(DeviceId.Type.OPEN_UDID);
        Countly.sharedInstance().init(config);
    }

    @Override
    protected void stopTrackingSession() {
        Countly.sharedInstance().halt();
    }

    @Override
    protected void trackingConfigChanged() {
        //As we cannot update configuration on the run, we need to restart the service
        stopTrackingSession();
        startTrackingSession();
    }

    @Override
    public void setUserIdentifier(String username) {
        if (isBugReportEnabled() || isAnalyticsEnabled()) {
            Map<String, String> userFields = new HashMap<>();
            userFields.put("username", username);
            Countly.userData.setUserData(userFields);
            Countly.userData.save();
        }
    }

    @Override
    public void trackViewOnStart(Activity activity) {
        Countly.sharedInstance().onStart(activity);
    }

    @Override
    public void trackViewOnStop(Activity activity) {
        Countly.sharedInstance().onStop();
    }

    @Override
    protected void logHandledException(Exception e) {
        Countly.sharedInstance().crashes().recordHandledException(e);
    }
}

File: app/src/main/java/org/digitalcampus/oppia/analytics/CountlyAnalytics.kt
package org.digitalcampus.oppia.analytics

import android.app.Activity
import android.app.Application
import android.content.Context
import ly.count.android.sdk.Countly
import ly.count.android.sdk.CountlyConfig
import ly.count.android.sdk.DeviceId
import org.digitalcampus.mobile.learning.BuildConfig

class CountlyAnalytics(appContext: Application) : BaseAnalytics(appContext) {
    override fun startTrackingSession() {
        val config = CountlyConfig(appContext, BuildConfig.COUNTLY_APP_KEY.toString(), BuildConfig.COUNTLY_SERVER_URL)
        config.setLoggingEnabled(Analytics.isTrackingEnabled(appContext))
        config.setViewTracking(Analytics.isTrackingEnabled(appContext))
        if (Analytics.isBugReportEnabled(appContext)) {
            config.enableCrashReporting()
        }
        config.setIdMode(DeviceId.Type.OPEN_UDID)
        Countly.sharedInstance().init(config)
    }

    override fun stopTrackingSession() {
        Countly.sharedInstance().halt()
    }

    override fun trackingConfigChanged() {
        //As we cannot update configuration on the run, we need to restart the service
        stopTrackingSession()
        startTrackingSession()
    }

    override fun setUserIdentifier(username: String) {
        if (isBugReportEnabled() || isAnalyticsEnabled()) {
            val userFields = mapOf("username" to username)
            Countly.userData.setUserData(userFields)
            Countly.userData.save()
        }
    }

    override fun trackViewOnStart(activity: Activity) {
        Countly.sharedInstance().onStart(activity)
    }

    override fun trackViewOnStop(activity: Activity) {
        Countly.sharedInstance().onStop()
    }

    override fun logHandledException(e: Exception) {
        Countly.sharedInstance().crashes().recordHandledException(e)
    }
}